### PR TITLE
#11642: demux ttnn::typecast into ttnn::experimental::typecast on gra…

### PR DESCRIPTION
**Pull request by sjameelTT**
- add typecast support to grayskull within ttnn::typecast by demuxing into ttnn::experimental::typecast
- merge typecast tests into test_copy_ops and add community test that didn't work on grayskulll
- add tests for grayskull

### Ticket
#11642 

### Problem description
Provide context for the problem.

### What's changed
Describe the approach used to solve the problem.
Summarize the changes made and its impact.

### Checklist
- [ ] Post commit CI passes
- [ ] Blackhole Post commit (if applicable)
- [ ] Model regression CI testing passes (if applicable)
- [ ] New/Existing tests provide coverage for changes

### CODE REVIEW

**Comment by sjameelTT** [comment]
https://github.com/tenstorrent/tt-metal/actions/runs/10589071578

File: ttnn/cpp/ttnn/operations/experimental/copy/typecast/typecast.cpp [hunk 1]
```diff
@@ -15,14 +15,16 @@ ttnn::Tensor TypecastOperation::invoke(
...
+    const std::optional<Tensor>& optional_output_tensor) {
+    return operation::run(ttnn::operations::data_movement::CopyDeviceOperation{output_mem_config.value_or(input_tensor.memory_config()), dtype}, {input_tensor}, {}, {optional_output_tensor}, queue_id).at(0);
 }
 
 ttnn::Tensor TypecastOperation::invoke(
     const Tensor& input_tensor,
     const DataType& dtype,
-    const std::optional<MemoryConfig>& output_mem_config) {
+    const std::optional<MemoryConfig>& output_mem_config,
+    const std::optional<Tensor>& optional_output_tensor) {
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/experimental/copy/typecast/typecast.cpp:27]: Don’t forget to pass optional output tensor here

File: ttnn/cpp/ttnn/operations/copy.hpp [hunk 2]
```diff
@@ -49,6 +50,15 @@ struct Typecast {
                 output_dtype == optional_output_tensor.value().get_dtype(),
                 "If both output dtype and output tensor provided dtype should match");
         }
+        if (input.device()->arch() == tt::ARCH::GRAYSKULL) {
+            TT_FATAL(
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/copy.hpp:54]: Should these fatals go to the experimental op?

Comment (sjameelTT) [reply ttnn/cpp/ttnn/operations/copy.hpp]: Yeah better to put them in the validates

File: ttnn/cpp/ttnn/operations/experimental/copy/typecast/typecast.cpp (hunk 1 above)
Comment (sjameelTT) [reply ttnn/cpp/ttnn/operations/experimental/copy/typecast/typecast.cpp:27]: Done

**Review by ayerofieiev-tt — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/experimental/copy/typecast/typecast.cpp (hunk 1 above)
Comment (Denys88) [reply ttnn/cpp/ttnn/operations/experimental/copy/typecast/typecast.cpp:27]: I think it should not be const?

**Review by ntarafdar — APPROVED** [review]
Great job @sjameelTT 

CC: @marty1885

File: ttnn/cpp/ttnn/operations/experimental/copy/typecast/typecast.cpp (hunk 1 above)
Comment (sjameelTT) [reply ttnn/cpp/ttnn/operations/experimental/copy/typecast/typecast.cpp:27]: The only thing we do with it is return by value in create_output_tensors and there's no direct modification...I'm not actually sure this feature works. Even if I drop the const here the infra actually requires an eventual const for some reason :/. We should probably migrate this to the new infra as part of a bigger refactor as that gets fixed.

**Pull request merged**